samples/dashboards/data/mqtt-connector/demo.js

```diff
@@ -0,0 +1,580 @@
+                }
+            },
+            timeout: 10
+        });
+    }
+
+    /**
+     * Unsubscribe from an MQTT topic.
+     *
+     */
+    unsubscribe() {
+        const { topic, subscribeEvent, failureEvent } = this.options;
+
+        this.mqtt.unsubscribe(topic, {
+            onSuccess: () => {
+                if (subscribeEvent) {
+                    subscribeEvent(false, topic);
+                }
+            },
+            onFailure: response => {
+                if (subscribeEvent) {
+                    failureEvent(response.errorMessage);
+                }
+            }
+        });
+
+        if (subscribeEvent) {
+            subscribeEvent(false, topic);
+        }
+    }
+
+    /**
+     * Process connection success
+     *
+     */
+    onConnect() {
+        this.connected = true;
+        const { host, port, user, connectEvent, autoSubscribe } = this.options;
+
+        if (connectEvent) {
+            connectEvent(true, host, port, user);
+        }
+
+        // Subscribe to the topic
+        if (autoSubscribe) {
+            this.subscribe();
+        }
+    }
+
+    /**
+     * Process incoming message
+     *
+     */
+    onMessageArrived(mqttPacket) {
+        // Executes in Paho.MQTT.Client context
+        const connector = connectorTable[this.clientId],
+            converter = connector.converter,
+            connTable = connector.table;
+
+        // Parse the message
+        const data = JSON.parse(mqttPacket.payloadString);
+        converter.parse({ data });
+        const convTable = converter.getTable();
+
+        // Append the incoming data to the current table
+        if (connTable.getRowCount() === 0) {
+            // First message, initialize columns for data storage
+            connTable.setColumns(convTable.getColumns());
+        } else {
+            // Subsequent message, append as row
+            connTable.setRows(convTable.getRows());
+        }
+
+        connector.packetCount++;
+
+        // Execute custom packet handler
+        const packetEvent = connector.options.packetEvent;
+        if (packetEvent) {
+            packetEvent(
+                mqttPacket.destinationName,
+                mqttPacket.payloadString,
+                connector.packetCount
+            );
+        }
+    }
+
+    /**
+     * Process lost connection
+     *
+     */
+    onConnectionLost(response) {
+        // Executes in Paho.MQTT.Client context
+        const connector = connectorTable[this.clientId];
+        connector.connected = false;
+        connector.packetCount = 0;
+
+        const { connectEvent, errorEvent } = connector.options;
+
+        // Execute custom connection handler
+        if (connectEvent) {
+            connectEvent(false);
+        }
+
+        if (response.errorCode !== 0) {
+            // Execute custom error handler
+            if (errorEvent) {
+                errorEvent(response.errorMessage);
+            }
```

Comment:
Dashboards API might have a design issue regarding events. In Highcharts the event callbacks are attached as event listeners and then event emitters in classes then will automatically inform all listeners including callbacks from options. Also event callbacks should have their own sub-object on options of type `Record<string, EventCallback>`. You should discuss this and have a consistent approach.

Reply:
I have rewritten this to use events.